core/pta/system.c

```diff
@@ -795,6 +796,168 @@ static TEE_Result system_dlsym(struct tee_ta_session *cs, uint32_t param_types,
+			alg == TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA1 ||
+			alg == TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1)
+			goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_MD5) && IS_ENABLED(CFG_CRYPTO_SHA1)) {
+			if (alg == TEE_ALG_RSASSA_PKCS1_V1_5_MD5SHA1)
+				goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_SHA224)) {
+		    if (alg == TEE_ALG_RSASSA_PKCS1_V1_5_SHA224 ||
+			alg == TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA224 ||
+			alg == TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA224)
+			goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_SHA256)) {
+		    if (alg == TEE_ALG_RSASSA_PKCS1_V1_5_SHA256 ||
+			alg == TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256 ||
+			alg == TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA256)
+			goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_SHA384)) {
+			if (alg == TEE_ALG_RSASSA_PKCS1_V1_5_SHA384 ||
+			    alg == TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA384 ||
+			    alg == TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA384)
+			goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_SHA512)) {
+			if (alg == TEE_ALG_RSASSA_PKCS1_V1_5_SHA512 ||
+			    alg == TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA512 ||
+			    alg == TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA512)
+			goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_RSASSA_NA1)) {
+			if (alg == TEE_ALG_RSASSA_PKCS1_V1_5)
+				goto check_element_none;
+		}
+		if (alg == TEE_ALG_RSA_NOPAD)
+			goto check_element_none;
+	}
+	if (IS_ENABLED(CFG_CRYPTO_DSA)) {
+		if (IS_ENABLED(CFG_CRYPTO_SHA1)) {
+			if (alg == TEE_ALG_DSA_SHA1)
+				goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_SHA224)) {
+			if (alg == TEE_ALG_DSA_SHA224)
+				goto check_element_none;
+		}
+		if (IS_ENABLED(CFG_CRYPTO_SHA256)) {
+			if (alg == TEE_ALG_DSA_SHA256)
+				goto check_element_none;
+		}
+
+	}
+	if (IS_ENABLED(CFG_CRYPTO_DH)) {
+		if (alg == TEE_ALG_DH_DERIVE_SHARED_SECRET)
+			goto check_element_none;
+	}
+	if (IS_ENABLED(CFG_CRYPTO_ECC)) {
+		if ((alg == TEE_ALG_ECDH_P192 || alg == TEE_ALG_ECDSA_P192) &&
+		    element == TEE_ECC_CURVE_NIST_P192)
+			return TEE_SUCCESS;
+		if ((alg == TEE_ALG_ECDH_P224 || alg == TEE_ALG_ECDSA_P224) &&
+		    element == TEE_ECC_CURVE_NIST_P224)
+			return TEE_SUCCESS;
+		if ((alg == TEE_ALG_ECDH_P256 || alg == TEE_ALG_ECDSA_P256) &&
+		    element == TEE_ECC_CURVE_NIST_P256)
+			return TEE_SUCCESS;
+		if ((alg == TEE_ALG_ECDH_P384 || alg == TEE_ALG_ECDSA_P384) &&
+		    element == TEE_ECC_CURVE_NIST_P384)
+			return TEE_SUCCESS;
+		if ((alg == TEE_ALG_ECDH_P521 || alg == TEE_ALG_ECDSA_P521) &&
+		    element == TEE_ECC_CURVE_NIST_P521)
+			return TEE_SUCCESS;
+	}
+	if (IS_ENABLED(CFG_CRYPTO_SM2_DSA)) {
+		if (alg == TEE_ALG_SM2_DSA_SM3 && element == TEE_ECC_CURVE_SM2)
+			return TEE_SUCCESS;
+	}
+	if (IS_ENABLED(CFG_CRYPTO_SM2_KEP)) {
+		if (alg == TEE_ALG_SM2_KEP && element == TEE_ECC_CURVE_SM2)
+			return TEE_SUCCESS;
+	}
+	if (IS_ENABLED(CFG_CRYPTO_SM2_PKE)) {
+		if (alg == TEE_ALG_SM2_PKE && element == TEE_ECC_CURVE_SM2)
+			return TEE_SUCCESS;
+	}
+
```

Comment:
What about a buig switch case:
```c
	switch (alg)
	case TEE_ALG_RSASSA_PKCS1_V1_5_MD5:
		if (IS_ENABLED(CFG_CRYPTO_RSA) && IS_ENABLED(CFG_CRYPTO_MD5))
			goto check_element_none;
		break;
	case TEE_ALG_RSASSA_PKCS1_V1_5_MD5:
		if (IS_ENABLED(CFG_CRYPTO_MD5))
			goto check_element_none;
		break;
	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA1:
	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA1:
	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1:
		if (IS_ENABLED(CFG_CRYPTO_SHA1))
			goto check_element_none;
		break;
	case TEE_ALG_RSASSA_PKCS1_V1_5_MD5SHA1:
		if (IS_ENABLED(CFG_CRYPTO_MD5) && IS_ENABLED(CFG_CRYPTO_SHA1))
				goto check_element_none;

	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA224:
	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA224:
	case TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA224:
		if (IS_ENABLED(CFG_CRYPTO_SHA224))
			goto check_element_none;
		break;
	(...)
	case TEE_ALG_ECDH_P192:
	case TEE_ALG_ECDSA_P192:
		if (IS_ENABLED(CFG_CRYPTO_ECC) &&
		    element == TEE_ECC_CURVE_NIST_P192)
			return TEE_SUCCESS;
		break;
	(...)
	case TEE_ALG_SM2_DSA_SM3:
		if (IS_ENABLED(CFG_CRYPTO_SM2_DSA) &&
		    element == TEE_ECC_CURVE_SM2)
			return TEE_SUCCESS;
		break;
	case TEE_ALG_SM2_KEP:
	if (IS_ENABLED(CFG_CRYPTO_SM2_KEP &&
		&& element == TEE_ECC_CURVE_SM2)
			return TEE_SUCCESS;
		break;
	case TEE_ALG_SM2_PKE:
	if (IS_ENABLED(CFG_CRYPTO_SM2_PKE))
		&& element == TEE_ECC_CURVE_SM2)
			return TEE_SUCCESS;
		break;
	case TEE_ALG_RSASSA_PKCS1_V1_5)
		if (IS_ENABLED(CFG_CRYPTO_RSASSA_NA1))
			goto check_element_none;
		break;
	case TEE_ALG_RSA_NOPAD)
		goto check_element_none;
	case TEE_ALG_DSA_SHA1:
		if (IS_ENABLED(CFG_CRYPTO_DSA) && IS_ENABLED(CFG_CRYPTO_SHA1))
			goto check_element_none;
		break;
	case TEE_ALG_DSA_SHA224:
		if (IS_ENABLED(CFG_CRYPTO_DSA) && IS_ENABLED(CFG_CRYPTO_SHA224))
				goto check_element_none;
		break;
	(...)
	default:
		break;
	}


	return TEE_ERROR_NOT_SUPPORTED;
check_element_none:
	if (element == TEE_CRYPTO_ELEMENT_NONE)
		return TEE_SUCCESS;
	return TEE_ERROR_NOT_SUPPORTED;
}

Reply:
That would work too, but I don't find it easier to read TBH.